src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -0,0 +1,150 @@
+        {
+            using (SafeX509StackHandle originatorCertStack = Interop.Crypto.NewX509Stack())
+            using (SafeX509StackHandle issuerRecipientCerts = Interop.Crypto.NewX509Stack())
+            using (SafeX509StackHandle skidRecipientCerts = Interop.Crypto.NewX509Stack())
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(issuerRecipientCerts);
+                Interop.Crypto.CheckValidOpenSslHandle(skidRecipientCerts);
+                Interop.Crypto.CheckValidOpenSslHandle(originatorCertStack);
+                                
+                ClassifyRecipients(recipients, issuerRecipientCerts, skidRecipientCerts);
+
+                foreach (X509Certificate2 cert in originatorCerts)
+                {
+                    using (SafeX509Handle certHandle = Interop.Crypto.X509Duplicate(cert.Handle))
+                    {
+                        if (!Interop.Crypto.PushX509StackField(originatorCertStack, certHandle))
+                            throw Interop.Crypto.CreateOpenSslCryptographicException();
+
+                        // The handle ownership has been transferred into the STACK_OF(X509).
+                        certHandle.SetHandleAsInvalid();
+                    }
+                }
+
+                int status;
+                using (SafeBioHandle contentBio = Interop.Crypto.CreateMemoryBio())
+                using (SafeAsn1ObjectHandle algoOid = Interop.Crypto.ObjTxt2Obj(contentEncryptionAlgorithm.Oid.Value))
+                {
+                    Interop.Crypto.CheckValidOpenSslHandle(algoOid);
+                    Interop.Crypto.CheckValidOpenSslHandle(contentBio);
+                    using (SafeCmsHandle cms = Interop.Crypto.CmsEncryptInitialize(algoOid, out status))
+                    {
+                        if (status == 0)
+                            throw new CryptographicException(SR.Cryptography_Cms_UnknownAlgorithm);
+
+                        CheckStatus(status);                   
+
+                        status = Interop.Crypto.CmsAddSkidRecipients(cms, skidRecipientCerts);
+                        CheckStatus(status);
+
+                        status = Interop.Crypto.CmsAddIssuerAndSerialRecipients(cms, issuerRecipientCerts);
+                        CheckStatus(status);
+
+                        status = Interop.Crypto.CmsAddOriginatorCerts(cms, originatorCertStack);
+                        CheckStatus(status);
+                        
+                        Interop.Crypto.BioWrite(contentBio, contentInfo.Content, contentInfo.Content.Length);
+                        status = Interop.Crypto.CmsEncryptFinalize(cms, contentBio);
+                        CheckStatus(status);
+
+                        byte[] encryptedMessage = Interop.Crypto.OpenSslEncode(
+                            Interop.Crypto.CmsGetEncodingLength,
+                            Interop.Crypto.CmsGetEncoding,
+                            cms);
```

Comment:
It is epsilon better to write this with lambdas instead of implicitly captured delegates:

```C#
byte[] encryptedMessage = Interop.Crypto.OpenSslEncode(
    handle => Interop.Crypto.CmsGetEncodingLength(handle),
    (handle, buf) => Interop.Crypto.CmsGetEncoding(handle, buf),
    cms);
```